### PR TITLE
Stabilize the `alloc` crate.

### LINKED ISSUE #27783: Tracking issue for location of facade crates
We probably don't want to indefinitely have a large number of facade crates which are all unstable, so we should eventually stabilize these crates, fold them back into the standard library, or find a good middle ground in which they can all reside. 

The current set of crates considered under this issue are:
- `rustc_unicode`
- `libc`
- `collections`
- `alloc`
- `rand`

----

**Update 2018-04-05:**

* `libc` and `rand` have moved to crates.io
* `collections` [was merged](https://github.com/rust-lang/rust/pull/42648) into `alloc`
* `rustc_unicode` was renamed to `std_unicode` and [later merged](https://github.com/rust-lang/rust/pull/49698) into `core`

This leaves only the `alloc` crate still unstable, which with the stable `core` and `std` crates (plus arguably `proc_macro`) form the standard library.

----

**Update 2018-06-19:**

[RFC 2480](https://github.com/rust-lang/rfcs/pull/2480) proposes stabilizing the `alloc` crate, which would close this issue.

### PR DESCRIPTION
**Pull request by SimonSapin**
This implements RFC 2480:

* https://github.com/rust-lang/rfcs/pull/2480
* https://github.com/rust-lang/rfcs/blob/master/text/2480-liballoc.md

Closes https://github.com/rust-lang/rust/issues/27783

### CODE REVIEW

**Comment by SimonSapin** [comment]
Note that the `alloc::prelude` module is not being stabilized in this PR. It was moved to a dedicated feature gate `alloc_prelude` and tracking issue https://github.com/rust-lang/rust/issues/58935.

**Comment by alexcrichton** [comment]
I think it's probably best to land this after the beta branch next week? Just want to give it a but more time to bake if we can.

Other than that though r=me!

**Comment by SimonSapin** [comment]
That’s fine if you prefer, though in a way this has been baking for years already. The only thing changing now is stability attributes.

**Comment by kennytm** [comment]
(Marking as blocked by the beta release)

File: src/liballoc/lib.rs [hunk 1]
```diff
@@ -51,10 +51,7 @@
 //! default global allocator. It is not compatible with the libc allocator API.
 
 #![allow(unused_attributes)]
-#![unstable(feature = "alloc",
-            reason = "this library is unlikely to be stabilized in its current \
-                      form or name",
-            issue = "27783")]
+#![stable(feature = "alloc", since = "1.35.0")]
```
Comment (Centril) [inline src/liballoc/lib.rs]:
```suggestion
#![stable(feature = "alloc", since = "1.36.0")]
```

**Comment by Centril** [comment]
@bors r-

Version number is wrong.

**Comment by SimonSapin** [comment]
Fixed.

@bors r=alexcrichton

**Pull request merged**